### ISSUE TITLE
LivePostion and LiveRange should be introduced

### ISSUE THREAD

**Issue by pjasiun** [ckeditor/ckeditor5-core#64]
`SmartPostion` is a position of in the document which changes the path when document change its structure. It listen on the document events and update every time document changes.

`SmartRange` is something more then pair of start and end `SmartPostion`. For example if the end position is moved before the start, the range should be collapsed.

`PositionIterator` also may need separate implementation for `SmartPosition`.

**Comment by fredck** [comment]
I'm not for calling them "smart". I think something like `LivePosition` and `LiveRange` or `DynamicPosition` and `DynamicRange` may work.

**Comment by pjasiun** [comment]
`Dynamic*` was my first idea so let's stick to it.

**Comment by Reinmar** [comment]
DOM often uses word "live". Live collections, live ranges, etc. And I think that this word best reflects what will happen with the range. It will stay alive while some document is changing.

**Comment by scofalik** [comment]
We should do a poll.

**Comment by fredck** [comment]
I proposed "live" as well, so I couldn't agree more :) btw, it is shorter and easier to write.

**Comment by scofalik** [comment]
It will be `Live` then. I am happy that you commented on it before I changed `Smart` to `Dynamic` :)

**Comment by scofalik** [comment]
I implemented `LivePosition`. It listens for `treeModel.Document` `change` event and gets transformed accordingly. It uses `stickiness` to decide how to behave when nodes are inserted at the same position as `LivePosition`. If `LivePosition` sticks to left, nodes are inserted "after" the `LivePosition` so it does not changes. If `LivePosition` sticks to right, nodes are inserted "before" it and the `LivePosition` is transformed.

For `LiveRange` I actually have three solutions. In all three we are also listening to `change` event.
t/99-c https://github.com/ckeditor/ckeditor5-core/tree/t/99-c
The first one. `LiveRange` does not use `LivePosition` as boundaries. It uses normal `Position` and transforms  them accordingly. Of course the first thing that comes to mind is that some of the job here is same as what we do in `LivePosition` transformation (i.e. we do exactly the same for `insert` change type).

t/99-d https://github.com/ckeditor/ckeditor5-core/tree/t/99-d
The second one. We use `LivePosition`s as boundaries. The code got a little shorter. No we don't just transform boundaries but in fact "fix" them if special cases happened. It's hard to tell if the code is really simpler as we have some special cases. Unfortunately we have to have a separate treatment if:
* ranges are equal: `LiveRange` becomes moved range.
* `LiveRange` is intersecting with moved range (but is not contained): special fix.
*  any other case (incl. `LiveRange` is contained in moved range): do nothing.
It's sad that we treat equal range differently because it really should be the same case as contained range but when you "moved-out" nodes from between two positions they are not moved as they are not in the range.

t/99-e https://github.com/ckeditor/ckeditor5-core/tree/t/99-e
The last one. We still use `LivePosition` but their behavior is changed. Now stickiness also is applied when you "move-out" nodes. So if you move out nodes from range `[ 0, 2 ]` to `[ 0, 5 ]` and you have `LivePosition` sticking to left on `[ 0, 5 ]` it also gets moved. This simplifies code in `LiveRange`. This seems like cleaniest solution but I don't know if this `LivePosition` behavior is intuitive and which behavior we/users will use/need more. If we go with this solution, I'd consider making stickiness a tri-state flag:
* STICKS_TO_LEFT - sticks to left for insertion and move-out
* STICKS_TO_RIGHT - sticks to right for insertion and move-out
* NOT_STICKY - in fact it sticks to right for insertion (because it has to end up on left or right) but is not sticky for move-out.
Or keep it as two-state flag but also make a flag for move-out stickiness.

Also I would like to consider stickiness for `Position` and re-view OT scripts because it might get useful there (I am not sure though).

**Comment by fredck** [comment]
> stickiness

As far as I remember, the only term I've ever heard for such thing is "gravity", which seems to sound better when it comes to insertion. In the other hand, "stickiness" plays well with the move-out thing. So I would propose have two different properties: "gravity" and "stickiness" (or maybe only "gravity" would be enough as it may endup that it will be always the same as "stickiness").

As for "left/right", there is a tendency to avoid such terminology to not make it limited to LTR. In RTL this should act in the opposite way. The neutral terms to replace them are "backward/forward", so this could be taken in consideration here as well.

**Comment by Comandeer** [comment]
Maybe we can use "start"/"end" instead of "left"/"right"? It's used in CSS properties that have to support LTR & RTL, e.g. `text-align: start` or `align-items: flexbox-start`.

**Comment by scofalik** [comment]
I think that before and after would be the best actually. Remember that we are talking about tree model, not the actual DOM or rendered view and in terms of tree model and nodes neighbourhood I think it is fine to say that something is after or before. Even in native DOM code there are words before and after used.

**Comment by scofalik** [comment]
There is one more thing that I don't like about that code, on all three branches. There is a method `.clone()` in `Position` and `Range`. The method should return a new instance of given object that has the same properties as the original object. `.clone()` is used in other methods of `Range` and `Position`. Both `.clone()` and those other methods are inherited by `LiveRange` and `LivePosition`. And that's great. The problem is that those methods also use `.clone()`. So whenever `LiveRange` uses inherited method that has `.clone()` the cloned object is also a `LiveRange`! This is important, because `LiveRange` binds events that need to be unbound! And sometimes those cloned objects "die" inside the inherited method so there is no place where one could use `.detach()` method. Even worse, `Range` does not have (and should not) `.detach()` method, so we can't detach those cloned objects inside inherited methods.

To prevent this, I changed how `.clone()` works for childclasses. Without parameter, it clones using parent class (`Range` or `Position`). When `true` is passed, it clones using child class. This way I don't have to change methods that use `.clone()`. But I think this looks and feels wrong. Also this may lead to unconscious creation of "live" objects without user realizing it. Which may lead to unbound events.

One of my ideas is to get rid of `.clone()` and introduce `Position.createFromPosition()` and `Range.createFromRange()`. Or even `Position.clone()` and `Range.clone()`. I'd then replace `this.clone()` uses into proper ones and the API user will have to explicitly state whether he wants to create "live" or normal object.

Do you have other ideas?

**Comment by pjasiun** [comment]
For me, `LivePosition.clone` with a parameter is fine. A little bit strange, but very useful. If one call `livePosition.clone()` this is unclear what one want to do: clone only the position or a position with events listeners. This is similar to the `object.clone( deep )`, so it is fine for me.

**Comment by pjasiun** [comment]
And the live position should definitely works as described in the `t/99-e`. Because, after https://github.com/ckeditor/ckeditor5-core/issues/62, it will be impossible to keep the reference to the text node, the way to keep it will be to put the live position before it with the right stickiness/gravity. Such range should have this text node always on the right, as long as it exists. Event if the node is moved the position should be moved with it. Otherwise it would be very hard to keep references to text nodes.

Also if you have `x | y` and move `y` and position does not move, it behave as it would be stick to the `x` not `y`.

**Comment by scofalik** [comment]
So we will go with solution that is on branch t/99-e https://github.com/ckeditor/ckeditor5-core/tree/t/99-e

I pushed some commits with further refactoring. I dropped `.clone()` method in favor of static creators `.createFromPosition` and `.createFromRange`. I also updates some things.

Still to do:
* Consider renaming stickiness
* Consider introducing stickiness for `Position`, not only for `LivePosition`
* If above is done, review OT scripts, maybe some will simplify.

**Comment by Reinmar** [comment]
Regarding gravity and stickiness, as far as we can tell (after a F2F discussion) these are two separate things now.

Stickiness describes how live positions behave. This is a property which is taken into consideration every time something changes next to this position.

Gravity, however, applies to the selection. It defines which styles selection will inherit – those on the left of it or on the right of it. It applies only once, when a new selection is created through the UI (by clicking, arrow keys, etc.).

**Comment by scofalik** [comment]
I am for keeping "sticky" name, as it is more intuitive what really happens. For now, after some discussion, I think that we are okay with having same behavior for insertion and moving, so I will keep two flags.
I am changing flag names to `STICK_TO_BEFORE` and `STICK_TO_AFTER`. If you come up with better names I'll consider changing them, but I think after and before are best in tree model context.
